### ISSUE TITLE
[Feature] CSS.forcePseudoState

### ISSUE THREAD

**Issue by gilles-yvetot** [microsoft/playwright#3347]
Hello there,

Thanks for the great work done on pptr and now on pw 👍 

I have been trying to force the CSS pseudo state of an element to hover. I want this css state to be "permanent" so I cannot use `elementHandle.hover()`. I have trying to use a CDP session to do that but it is very hard to find the `nodeId` needed to call this function. Especially when you work with extension iframes (I could not find a way to identify an element by its XPath).

Thanks a lot

**Comment by gilles-yvetot** [comment]
One way I could solve the issue is to have a way to access the CDP client/session of a frame or an elementHandle

**Comment by mxschmitt** [comment]
This issue looks pretty similar to #3345

**Comment by gilles-yvetot** [comment]
Yeah they look very similar. Sorry I did not see it before posting

**Comment by pavelfeldman** [comment]
Could you share more on your use case?

**Comment by gilles-yvetot** [comment]
The idea is to let a user select an element in the page but I'd like to keep it "activated". And some elements only stay visible on hover.

**Comment by yury-s** [comment]
@gilles-yvetot do I get it right that you want to "activate" the pseudo element, keep it in that state and and the let the user select it? If so, can you clarify what is the user action that would "select" the pseudo-element?

**Comment by gilles-yvetot** [comment]
@yury-s I want to activate the CSS pseudo state, like `hover`, not the pseudo element, like `:before`.

But the real question is more to be able to access the CDP client/session of an elementHandle to retrieve the `nodeId` that I can use with CDP after

**Comment by yury-s** [comment]
You could create CDP session for the page and then evaluate node id using CDP commands, not much help from playwright in that case but it should solve your problem.

**Comment by gilles-yvetot** [comment]
@yury-s Yep that's what I ended up doing. But I could not use XPath though. So I converted my XPath to a classic query selector then start forcing `hover` from the document root to the target element (if that makes sense)

**Comment by dgozman** [comment]
@gilles-yvetot Did you manage to find `nodeId` with your workaround? Overall, this does not seem like a feature that would be useful to a lot of Playwright users, but I'd like to help you with your CDP quest 😄

**Comment by gilles-yvetot** [comment]
@dgozman thanks a lot! Yeah I have a working solution:
- converting the Path to a selector
- enabling CSS and DOM
- getting the document
- forcing the CSS state `hover` on the root, aka `<html>` element
- adding "parts" of the selector to get the child of the root, e.g `<body>`
- forcing the CSS state on this child element
- keep adding "parts" of the selector and forcing the state until the desired node

**Comment by dgozman** [comment]
@gilles-yvetot Wonderful! Feel free to ask more questions when the help is needed.

**Comment by codingedgar** [comment]
> @dgozman thanks a lot! Yeah I have a working solution:
> 
> * converting the Path to a selector
> * enabling CSS and DOM
> * getting the document
> * forcing the CSS state `hover` on the root, aka `<html>` element
> * adding "parts" of the selector to get the child of the root, e.g `<body>`
> * forcing the CSS state on this child element
> * keep adding "parts" of the selector and forcing the state until the desired node

How did you do it? could you share a snippet?

**Comment by jonathantneal** [comment]
For posterity, I had to do something similar, and this is what it looked like in code.

**Converting the Path to a selector**:
Here is a [gist of the logic I used to get the NodeID of an Element](https://gist.github.com/jonathantneal/1b25514f6120254b152d1232896a9f1e), which in my case needed to cross iframe boundaries.

**Enabling CSS and DOM**:
```ts
await cdp().send("DOM.enable")
await cdp().send("CSS.enable")
```

From here, I skip to forcing the pseudo class. I did not need to get the document, as I did not desire toggling `&:hover` down the entire tree. This is how I forced the `hover` pseudo class.

```ts
const nodeId = await getNodeId(element)

await cdp().send("CSS.forcePseudoState", { nodeId, forcedPseudoClasses: ["hover"] })
```
